[LC-283] Add basic integration test

metalg0su opened the pull request:
# Test steps:
To run tests, follow step below...
1. clone the repository and make virtualenv 
2. `$ make develop`
3. `$ pytest testcase/integration [-v] [options]`
> **_NOTE_**: Possible [options] are:
> - `--peer-count=[int]`: execute test with [int] peers
> - `--channel-count=[int]`: execute test with [int] channels

**_... or you can simply run tests by `$ make test`_**

# Test check list
This test tries to...
- Run loopchain launcher and check stdout lines to find any exception raised
- Check all peers and channels are ready to receive transactions 
- Send transaction and get valid tx_hash
- Check Sent transactions are successfully synced between peers (also channels)

# Test script
- `conftest.py`: contains common test fixture and helper for making loopchain config.
- `test_for_fixtures.py`: contains basic tests to be run before integration test.
- `test_integration.py`: contains tests for loopchain integration scenario. 

# Note
- Ensure that there are no loopchain-related process is running!

metalg0su committed:
[LC-283] Edit test related files

 - `make develop`

Co-authored-by: June Park <laurenceyoon@gmail.com>

metalg0su committed:
[LC-283] Add basic integration test

Co-authored-by: June Park <laurenceyoon@gmail.com>

metalg0su commented on testcase/integration/test_integration.py:
```diff
@@ -0,0 +1,327 @@
+                .nid(3) \
+                .nonce(100) \
+                .data(msg) \
+                .build()
```
Thank you for your suggestion.

How about this one?
I found that iconsdk provides "MessageTransactionBuilder.from_dict()" method.

```
            transaction_data = {
                "from": from_to_address,
                "to": from_to_address,
                "step_limit": 100000000,
                "nid": 3,
                "nonce": 100,
                "data": msg,
            }
            transaction = MessageTransactionBuilder().from_dict(transaction_data).build()
            signed_transaction = SignedTransaction(transaction, wallet)
```

or simply...
```
transaction_data: dict = { 
    "from": ...
    ...
}
transaction = MessageTransactionBuilder(**transaction_data)
...
```

metalg0su commented on testcase/integration/test_integration.py:
```diff
@@ -0,0 +1,327 @@
+            tx_hash = icon_service.send_transaction(signed_transaction)
+            print("Tx hash: ", tx_hash)
+
+            assert tx_hash[:2] == "0x"
```
Thank you!

metalg0su commented on testcase/integration/test_integration.py:
```diff
@@ -0,0 +1,327 @@
+        assert tx_list["nid"] == expected_data["nid"]
+        assert tx_list["accounts"] == expected_data["accounts"]
+
+    @pytest.mark.skip(reason="Could be a redundent test")
```
Thank you!

metalg0su commented on testcase/integration/conftest.py:
```diff
@@ -0,0 +1,304 @@
+                "\"AnD ... and ...and...\""
+                "THis conJunction carriEs enouGh force to shaKe and uproot the verb \"to be.\" "
+                "Where are You goIng? Where are you coMing from? What are you heading for? "
+                "These are totally useless questions.\n\n- 『Mille Plateaux』, Gilles Deleuze & Felix Guattari\n\n\""
```
Fixed it.
Thank you!

metalg0su commented on testcase/integration/test_integration.py:
```diff
@@ -0,0 +1,327 @@
+
+        Assertion Tests:
+            1. Queried balance == Expected balance
+            2. All queired balance has same value
```
Fixed it.
Thank you!

metalg0su commented on testcase/integration/conftest.py:
```diff
@@ -0,0 +1,304 @@
+        with open(config_path) as f:
+            each_config = json.load(f)
+
+        for channel_name in each_config["CHANNEL_OPTION"].keys():
```
Thank you!

metalg0su commented on testcase/integration/conftest.py:
```diff
@@ -0,0 +1,304 @@
+        }
+        peers.append(each_peer)
+
+    for channel_name in channel_manage_data.keys():
```
Thank you!

metalg0su commented on testcase/integration/conftest.py:
```diff
@@ -0,0 +1,304 @@
+                "balance": "0x0"
+            }
+        ]
+        genesis_path = str(tmp_path / "genesis_test.json")
```
I changed all-related lines in the code to below:
`os.path.join(tmp_path, "file_name.here")`

os.path.join() accepts os.PathLike protocol, which means `pathlib` acceptable.
https://docs.python.org/3/glossary.html#term-path-like-object.

Waiting your valuable opinions.

laurenceyoon commented on testcase/integration/test_integration.py:
```diff
@@ -0,0 +1,278 @@
+            port = 9000 + (peer_order * 100)
+
+            for channel_num in range(channel_count):
+                url = f"http://localhost:{port}/api/v3/channel_{channel_num}"
```
You can replace it with the method in utils. (Not only this, but also all parts of making url manually.)

```py
channel_name = f"channel_{channel_num}"
url = utils.normalize_request_url(port, conf.ApiVersion.v3, channel_name)
```

metalg0su replied:
`normalized_request_url` is useful. Thank you.

laurenceyoon commented on testcase/integration/conftest.py:
```diff
@@ -0,0 +1,268 @@
+    with open(target_peer_conf_path) as f:
+        default_peer_conf = json.load(f)
+
+    return default_peer_conf
```
I think it's confusing to return `default_peer_conf` when method_name is generally getting peer_info. If its usage is just for the default peer info, the method name should be changed instead.

metalg0su replied:
I agree that variable names used in those line raise terrible confusion. So I changed.

I intended that the function `get_peer_info` helps extracting `order`'s peer information in given list. It means, the expected return value is should be peer info, as dict type.

So, I changed confusing variable `default_peer_conf` to simply `peer_info`

laurenceyoon commented on testcase/integration/test_integration.py:
```diff
@@ -0,0 +1,278 @@
+        peer_count = int(request.config.getoption("--peer-count"))
+        channel_count = int(request.config.getoption("--channel-count"))
+
+        billionaire = genesis_data["transaction_data"]["accounts"][2]
```
I guess `genesis_node` would be fine.

metalg0su replied:
Thank you for your helpful suggestion!

metalg0su commented on setup.py:
```diff
@@ -20,6 +20,9 @@
     'license': "Apache License 2.0",
     'install_requires': requires,
+    'extras_require': {
+        'tests': ['iconsdk==1.0.8', 'pytest>=4.6.3', 'pytest-xprocess>=0.12.1'],
```
good news! Thank you

metalg0su commented on testcase/integration/test_for_fixtures.py:
```diff
@@ -0,0 +1,89 @@
+import os
+import time
+
+import conftest
```
Thank you. I learned how to import from relative path at this time.

laurenceyoon commented on .gitignore:
```diff
@@ -236,6 +236,7 @@ virtualenv-15.2.0-py2.py3-none-any.whl
 *.whl
 
 user.mk
+.xprocess/
```
How about add removing `.hypothesis/`, `.xprocess/`, `.pytest_cache/` directory when we run `make clean`?

metalg0su replied:
Thank you
Changed:
- Add above folders to `.gitignore`
- Add `$ make cleanf-cache` to remove them (Added this keyword to `$ make clean`

metalg0su commented on testcase/integration/test_integration.py:
```diff
@@ -0,0 +1,281 @@
+            # Create message
+            byte_msg = f"test_msg on {random.randint(0, 44444)}".encode("utf-8")
+            hex_msg = b"0x" + binascii.hexlify(byte_msg)
+            msg = hex_msg.decode("utf-8")
```
Thank you !

metalg0su commented on testcase/integration/test_integration.py:
```diff
@@ -0,0 +1,281 @@
+
+            await_sec = 0.5 * len(peer_conf_path_list)
+            print(f"Await consensus...({await_sec})")
+            time.sleep(await_sec)
```
hmm.. I encountered a problem that a consensus is delayed when I tried to test in case of numerous channels and a quite large number of peers.

I understand that running multiple peers in poor machine could end to undesirable result. So  I decided to wait for a moment in every request to ensure consensus.

metalg0su commented on testcase/integration/test_integration.py:
```diff
@@ -0,0 +1,281 @@
+            port = 9000 + (peer_order * 100)
+
+            for channel_num in range(channel_count):
+                url = utils.normalize_request_url(f"http://localhost:{port}", conf.ApiVersion.v3, f"channel_{channel_num}")
```
Thank you. 
I changed all related lines to use `normalize_request_url`

metalg0su committed:
[LC-283] Additional changes

 - Use block verifier and serializer to deal with responsed block dict
 - Split channel & port loop from test logic

metalg0su committed:
[LC-283] Change docstring type to rst

yakkle commented on Makefile:
```diff
@@ -100,6 +105,12 @@ clean-log:
 	@echo "Cleaning up logs..."
 	@rm -rf log/
 
+clean-cache:
```
I think `clean-test` is better. It' just suggestion.

metalg0su replied:
Approved!

yakkle commented on testcase/integration/test_for_fixtures.py:
```diff
@@ -0,0 +1,87 @@
+        with open(wallet_key_path) as f:
+            content = json.load(f)
+
+        print("KEY_CONTENT: ", content)
```
Is there alternative option instead of `print`?

yakkle reviewed (approved):
looks good! 👍
